src/psyclone/core/signature.py

```diff
@@ -99,6 +105,58 @@ def __getitem__(self, indx):
     def __str__(self):
         return "%".join(self._signature)
 
+    # ------------------------------------------------------------------------
+    def to_fortran(self, component_indices):
```

Comment:
OK, that might need some discussion. I would argue that you should get the terms that you provided in any error message. ATM we only allow Fortran as input language. And I don't see any way close to being as concise as what I have implemented.
My suggestion: we rename this function to (say) "to_input_language". We can then later, e.g. when we have other input language, query the language in the above function, and create output appropriate to the language currently used.

Reply:
In his work on adjointificaterisation, @rupertford has come up with a solution to this which at first I didn't like but I think does actually make sense: he passes an instance of a Visitor to the class and then, whenever he needs to generate a 'nice' error message, it's available to convert PSyIR into code. That way the class does not need to know what language is being used and we can still get nice error messages. If `Signature` kept a `Reference` (probably #1320 again?) then it would be really easy to do `my_visitor(signature.reference)` when required.

Reply:
OK, I will need to think about this, sounds like a good idea. But I should clarify one thing: I cannot store a Reference with the Signature (without significant refactoring). Simple example:
```
   a(i)%b(j) = 2
   a(j)%b(i) = 3
```
A VariablesAccessInfo instance will store both accesses for the same signature `(a,b)`, which is the key to get all accesses,. But I obviously cannot store a Reference with the signature here (which one should it be??) The Reference IS stored with each access. So e.g.:
```
all_access_info[Signature( (a,b) )] [0].node  --> a(i)%b(j)    # first access, assigning 2
all_access_info[Signature( (a,b) )] [1].node  --> a(j)%b(i)    # second access, assigning 3
```
So, once #1320 is done, I can switch to use the node information, but till that is done, I don't see any way to do that. But I can combine the signature with the ComponentIndices information to re-create the string that the psyir node will eventually produce.

Reply:
That indeed needs to be postponed till #1320, since at this stage we still might have strings, which a visitor cannot handle.  But I have added one TODO #1320 in this function.
*grin* I can't help myself from pointing out that psyGen.py also has a `bool_to_fortran` function, dynamo0p3 two `..._bound_fortran` functions as well ;)

Reply:
True, but they pre-date (I think/hope!) our move towards PSyIR and this is new code :-)